[Enhancement] Add build verification workflow rule with log polling instead of synchronous blocking

## What

Add a workflow rule that defines a non-blocking build verification pattern: run builds in background and poll logs for results, instead of synchronously waiting for build completion.

### Current Behavior
- When Claude Code executes a build command via the `Bash` tool, it blocks synchronously until the command completes or hits the timeout (default 120s, max 600s)
- Long-running builds appear to "hang" with no feedback to the user
- If the build exceeds the timeout, the command fails without useful error information
- The `_policy.md` states "All builds must pass before PR" but defines no specific method for build verification
- The `ci-debugging` skill provides manual diagnostic commands (`gh run view --log-failed`) but these are not integrated into a build verification workflow

### Expected Behavior
- Build commands are executed with `run_in_background: true` for long-running builds
- Build logs are periodically checked using `TaskOutput` or `tail` to detect success/failure early
- Clear workflow steps: launch build → poll logs → detect result → report or fix
- Timeout-aware: different strategies for short builds (< 30s, run inline) vs long builds (> 30s, background + poll)
- Integrates with `ci-debugging` skill for failure diagnosis

### Scope
- 1 new workflow rule file
- Update to `_policy.md` to reference the new rule

## Why

### Problem Statement

Claude Code's default behavior for build commands is synchronous blocking. For builds that take 30+ seconds (C++ compilation, large test suites, CI pipeline checks), this creates a poor experience:

1. **Perceived hang**: No output appears while waiting, making it seem frozen
2. **Timeout waste**: If a build fails at minute 3 of a 10-minute build, the error is only visible after the full timeout
3. **No early detection**: Build failures producing error output early in the log are not caught until completion
4. **Lost context**: When a build times out, the partial output may be truncated

The `Bash` tool supports `run_in_background: true` and `TaskOutput` for non-blocking execution, but no rule guides Claude Code to use this pattern for builds.

### Impact
- **Medium-High** - Affects every project with non-trivial build steps
- Build verification is a daily workflow for developers
- Current approach wastes time on failed builds

### Business Value
- **Faster feedback**: Detect build failures early via log polling
- **Better UX**: Users see progress instead of a hanging terminal
- **Smarter debugging**: Failed builds immediately trigger `ci-debugging` diagnosis flow
- **Resource efficiency**: No wasted wait time on already-failed builds

### Priority Justification
**Phase 2 (Short-term)** - Uses existing Bash tool capabilities, only needs workflow guidance

## Who

### Stakeholders
- All developers using Claude Code for build/test tasks
- CI/CD pipeline users
- Users working with compiled languages (C++, Rust, Java, Go)

## When

### Timeline
- Implementation: 1-2 hours
- Testing: 30 minutes

## Where

### Files to be Modified

```
project/.claude/rules/workflow/build-verification.md   📝 CREATE
global/commands/_policy.md                             📝 UPDATE (add reference)
```

## How

### Technical Approach

#### New Rule: `build-verification.md`

```markdown
---
paths:
  - "**/CMakeLists.txt"
  - "**/Makefile"
  - "**/package.json"
  - "**/build.gradle*"
  - "**/Cargo.toml"
  - "**/pyproject.toml"
  - "**/go.mod"
  - "**/*.csproj"
alwaysApply: false
---

# Build Verification Workflow

## Strategy Selection

| Build Duration Estimate | Strategy | Method |
|------------------------|----------|--------|
| < 30 seconds | Inline | Run synchronously, read output directly |
| 30s - 5 minutes | Background + Poll | `run_in_background: true` + `TaskOutput` polling |
| > 5 minutes (CI) | CI Log Check | `gh run view --log-failed` after trigger |

## Background Build Pattern

For builds expected to take > 30 seconds:

### Step 1: Launch in Background
Execute the build command with `run_in_background: true`:

    Bash(command="cmake --build build/ 2>&1", run_in_background=true)
    # Returns task_id for later polling

### Step 2: Poll for Results
Check build progress periodically using TaskOutput:

    TaskOutput(task_id="<id>", block=false)
    # Non-blocking check of current output

### Step 3: Detect Outcome
Scan output for success/failure indicators:

| Language | Success Indicator | Failure Indicator |
|----------|------------------|-------------------|
| C++/CMake | `Build complete` | `error:`, `fatal error:` |
| Node.js | `Successfully compiled` | `ERROR in`, `Failed to compile` |
| Gradle | `BUILD SUCCESSFUL` | `BUILD FAILED` |
| Rust | `Finished` | `error[E` |
| Go | Clean exit | `cannot find`, `undefined:` |
| Python | Clean exit | `SyntaxError`, `ImportError` |

### Step 4: On Failure → Diagnose
If failure detected, invoke ci-debugging diagnostic flow:
1. Categorize failure type (environment, platform, dependency, code)
2. Read relevant error lines from build log
3. Propose fix based on error pattern

## CI Pipeline Verification

For GitHub Actions / remote CI:

### Trigger and Check
    # After push, check CI status
    gh run list --limit 1 --json status,conclusion
    
    # If failed, get failure logs
    gh run view --log-failed

### Do NOT
- Wait indefinitely for CI completion
- Poll CI status in a tight loop (respect API rate limits)
- Assume CI passed without checking

## Integration with ci-debugging Skill

When build failure is detected, follow the diagnostic flow from the
ci-debugging skill:

1. Environment/Auth → Platform → Dependencies → Code Bug
2. Use appropriate fallback strategies
3. Report findings with specific error lines
```

#### Update `_policy.md`

Add reference to the new build verification rule:

```markdown
## Build Verification

| Stage | Requirement |
|-------|-------------|
| Before PR | All builds must pass |
| Before merge | All CI checks must pass |
| Exception | Draft PRs for work-in-progress |
| **Method** | **See `workflow/build-verification.md` for verification patterns** |
```

### Acceptance Criteria
- [ ] Build verification rule created with YAML frontmatter for build-file paths
- [ ] Three strategies defined: inline, background+poll, CI log check
- [ ] Success/failure indicators documented per language
- [ ] Integration with ci-debugging skill referenced
- [ ] `_policy.md` updated with reference to new rule
- [ ] Pattern tested with a sample build command

### Breaking Changes
None - New rule file and minor policy update

### Rollback Plan
1. Remove `project/.claude/rules/workflow/build-verification.md`
2. Revert `_policy.md` changes

## Additional Context

### Current Build-Related Configuration

| File | Role | Gap |
|------|------|-----|
| `_policy.md` | "Builds must pass" policy | No method specified |
| `build.md` | Dependency management guide | No verification workflow |
| `ci-debugging/SKILL.md` | CI failure diagnosis | Manual invocation only |
| `validate-skills.yml` | GitHub Actions workflow | Only validates SKILL.md files |

### Bash Tool Capabilities (Already Available)

| Feature | Parameter | Purpose |
|---------|-----------|---------|
| Background execution | `run_in_background: true` | Non-blocking build |
| Output retrieval | `TaskOutput(task_id, block=false)` | Poll build logs |
| Timeout control | `timeout: 600000` | Up to 10 minutes |

### Related Issues
- #120 (CI debugging skill) - Diagnostic flow for build failures
- #112 (New hook types) - Async hooks for non-blocking operations

--- a/global/commands/_policy.md
+++ b/global/commands/_policy.md
@@ -40,3 +40,4 @@ These policies apply to all commands in this directory.
 | Before PR | All builds must pass |
 | Before merge | All CI checks must pass |
 | Exception | Draft PRs for work-in-progress |
+| Method | See `rules/workflow/build-verification.md` for verification patterns |

--- a/project/.claude/rules/workflow/build-verification.md
+++ b/project/.claude/rules/workflow/build-verification.md
@@ -1,0 +1,149 @@
+---
+paths:
+  - "**/CMakeLists.txt"
+  - "**/*.cmake"
+  - "**/Makefile"
+  - "**/makefile"
+  - "**/package.json"
+  - "**/build.gradle*"
+  - "**/settings.gradle*"
+  - "**/pom.xml"
+  - "**/Cargo.toml"
+  - "**/pyproject.toml"
+  - "**/go.mod"
+  - "**/*.csproj"
+  - "**/*.sln"
+alwaysApply: false
+---
+
+# Build Verification Workflow
+
+## Strategy Selection
+
+Choose strategy based on expected build duration:
+
+| Build Duration Estimate | Strategy | Method |
+|------------------------|----------|--------|
+| < 30 seconds | Inline | Run synchronously, read output directly |
+| 30s - 5 minutes | Background + Poll | `run_in_background: true` + `TaskOutput` polling |
+| > 5 minutes (CI) | CI Log Check | `gh run view --log-failed` after trigger |
+
+### Duration Estimates by Build System
+
+| Build System | Typical Duration | Default Strategy |
+|-------------|-----------------|------------------|
+| `go build` / `go test` | < 30s | Inline |
+| `cargo check` | < 30s | Inline |
+| `npm run build` (small) | < 30s | Inline |
+| `cmake --build` | 30s - 5min | Background + Poll |
+| `gradle build` | 30s - 5min | Background + Poll |
+| `cargo build` (full) | 30s - 5min | Background + Poll |
+| `npm run build` (large) | 30s - 5min | Background + Poll |
+| Full test suites | 1 - 10min | Background + Poll |
+| GitHub Actions CI | 5min+ | CI Log Check |
+
+## Inline Strategy
+
+For builds expected under 30 seconds, run synchronously:
+
+```
+Bash(command="go build ./...", timeout=60000)
+Bash(command="cargo check", timeout=60000)
+```
+
+Read output directly. If it fails, diagnose from the error output.
+
+## Background + Poll Strategy
+
+For builds expected over 30 seconds:
+
+### Step 1: Launch in Background
+
+```
+Bash(command="cmake --build build/ --config Release 2>&1", run_in_background=true)
+# Returns task_id for later polling
+```
+
+### Step 2: Poll for Results
+
+Check build progress periodically using non-blocking reads:
+
+```
+TaskOutput(task_id="<id>", block=false, timeout=10000)
+# Returns current output without waiting for completion
+```
+
+Poll at reasonable intervals (every 10-15 seconds). Do NOT poll in a tight loop.
+
+### Step 3: Detect Outcome
+
+Scan output for success/failure indicators:
+
+| Language | Success Indicator | Failure Indicator |
+|----------|------------------|-------------------|
+| C++/CMake | `Built target`, clean exit | `error:`, `fatal error:` |
+| Node.js | `Successfully compiled`, `✓` | `ERROR in`, `Failed to compile` |
+| Gradle | `BUILD SUCCESSFUL` | `BUILD FAILED` |
+| Rust | `Finished` | `error[E` |
+| Go | Clean exit (no output) | `cannot find`, `undefined:` |
+| Python | Clean exit | `SyntaxError`, `ImportError`, `ModuleNotFoundError` |
+| C# | `Build succeeded` | `Build FAILED`, `error CS` |
+
+### Step 4: On Failure — Diagnose
+
+When failure is detected:
+
+1. Read the error lines from build output
+2. Categorize: compile error, linker error, missing dependency, test failure
+3. Apply fix based on error pattern
+4. Re-run build to verify fix
+
+Do NOT retry the same build without making changes — diagnose first.
+
+## CI Pipeline Verification
+
+For remote CI (GitHub Actions):
+
+### Trigger and Check
+
+```bash
+# After push, wait briefly for workflow to register
+sleep 5
+
+# Get latest run ID
+gh run list --branch <branch> --limit 1 --json databaseId -q '.[0].databaseId'
+
+# Poll CI status (non-blocking)
+gh run view <RUN_ID> --json status,conclusion -q '{status: .status, conclusion: .conclusion}'
+```
+
+### Interpret Result
+
+| status | conclusion | Action |
+|--------|-----------|--------|
+| completed | success | Proceed |
+| completed | failure | Fetch logs: `gh run view <RUN_ID> --log-failed` |
+| in_progress | — | Poll again after 30s |
+| queued | — | Poll again after 30s |
+
+### Rate Limiting
+
+- Poll CI status no more than once per 30 seconds
+- Maximum poll duration: 10 minutes
+- After timeout: report last known status, do not block indefinitely
+
+## Integration with ci-debugging Skill
+
+When build or CI failure is detected, follow the ci-debugging diagnostic flow:
+
+1. **Environment/Auth** — TLS errors, token expired, permission denied
+2. **Platform-Specific** — Works on one OS but fails on another
+3. **Missing Dependencies** — Module not found, package missing
+4. **Actual Code Bug** — Test assertions, logic errors
+
+## Rules
+
+- **Never wait indefinitely** for build or CI completion
+- **Never poll in a tight loop** — minimum 10s interval for local, 30s for CI
+- **Never assume success** without checking output or exit code
+- **Always diagnose before retrying** — blind retries waste time

